velox/expression/Expr.h

```diff
@@ -63,25 +63,64 @@ struct ExprStats {
   }
 };
 
+class ExprSaver {
+ public:
+  /// Persist the passed in 'RowVector' on disk. The data will be persisted at
+  /// $basePath/velox_vector_xxx that will be returned by storedPath.
+  /// Returns true if persistence is successful, false otherwise. In the case of
+  /// unsuccessful persistence, the reason will be returned by failMsg.
+  static bool saveData(
+      const BaseVector* FOLLY_NONNULL vector,
+      const char* FOLLY_NONNULL basePath,
+      std::string& storedPath,
+      std::string& failMsg);
+
+  /// Persist the passed in 'Expr' on disk in the form of sql. The sql will be
+  /// persisted at $basePath/velox_sql_xxx that will be returned by storedPath.
+  /// Returns true if persistence is successful, false otherwise. In the case of
+  /// unsuccessful persistence, the reason will be returned by failMsg.
+  static bool saveSql(
+      const std::string& sql,
+      const char* FOLLY_NONNULL basePath,
+      std::string& storedPath,
+      std::string& failMsg);
+};
+
+class ExprRestorer {
```

Comment:
Do we need to couple the logic of reading vector and SQL from files with expression evaluation?

Reply:
I could not find a better place to put these saver and restorer. They are too small to be put in a separate file..

Reply:
Can they go into VectorSaver?

Reply:
It deals with expressions as well which is technically not vector. The reason I put it here is kind of natural because in ExprExceptionContext it has both data and sql. Maybe we can put all ExprExceptionContext, the saver and restorer in a separate file, maybe named ExprUtil.h?